clang/lib/AST/ExprConstant.cpp

```diff
@@ -9125,9 +9126,25 @@ bool
 LValueExprEvaluator::VisitCompoundLiteralExpr(const CompoundLiteralExpr *E) {
   assert((!Info.getLangOpts().CPlusPlus || E->isFileScope()) &&
          "lvalue compound literal in c++?");
-  // Defer visiting the literal until the lvalue-to-rvalue conversion. We can
-  // only see this when folding in C, so there's no standard to follow here.
-  return Success(E);
+  APValue *Lit;
+  // If CompountLiteral has static storage, its value can be used outside
+  // this expression. So evaluate it once and store it in ASTContext.
+  if (E->hasStaticStorage()) {
+    Lit = E->getOrCreateStaticValue(Info.Ctx);
+    Result.set(E);
+    // Reset any previously evaluated state, otherwise evaluation below might
+    // fail.
+    // FIXME: Should we just re-use the previously evaluated value instead?
+    *Lit = APValue();
+  } else {
+    Lit = &Info.CurrentCall->createTemporary(E, E->getInitializer()->getType(),
+                                             ScopeKind::FullExpression, Result);
```

Comment:
Reading through the specification: https://gcc.gnu.org/onlinedocs/gcc/Compound-Literals.html
`ScopeKind::FullExpression` is correct for C++, but it should be `ScopeKind::Block` for C.

Given the assertion above `assert((!Info.getLangOpts().CPlusPlus || E->isFileScope())`, I guess we can only get here in C? So the correct value is `ScopeKind::Block` + an assertion that we never get here in C++?